lang/c++/impl/Compiler.cc

```diff
@@ -423,8 +423,12 @@ static NodePtr makeFixedNode(const Entity &e,
 static NodePtr makeArrayNode(const Entity &e, const Object &m,
                              SymbolTable &st, const string &ns) {
     auto it = findField(e, m, "items");
-    NodePtr node = NodePtr(std::make_shared<NodeArray>(
-        asSingleAttribute(makeNode(it->second, st, ns))));
+    std::shared_ptr<NodeArray> nodePointer = std::make_shared<NodeArray>(
+        asSingleAttribute(makeNode(it->second, st, ns)));
+    NodePtr node = NodePtr(nodePointer);
+    if (containsField(m, "element-id")) {
+        nodePointer->elementId_ = getStringField(e, m, "element-id");
+    }
```

Comment:
I'm kind of confused about what element id is. In my head, element-id was a numeric label indicating the position in the array. Is that not the case? Looking at some other impl, it does look like it's an int, but it's possible that may be an internal id that indexes the actual string value.

https://github.com/Netflix/iceberg/blob/5684ac1b77851737ed316e7f9d302f902b3e1e67/core/src/main/java/com/netflix/iceberg/avro/AvroSchemaUtil.java#L215-L219

Do you have any pointers to more context about element-ids?

Reply:
I tried googling it too, and couldn't find anything terribly useful. A lot of this stuff seems undocumented. I'm primarily working backwards from a schema generated by pyspark. I wrote a script to extract the schema from valid manifest and manifest list files. I tried removing things to see what's actually required. `element-id` is required here by both the Java and Python versions.

Since it's a value in the schema, my guess is that it's meant to identify the array element type, in the same way that `field-id` is used for record fields. So if you change the element type of an array without changing the name of an array, the system will be able to detect the schema evolution and parse the elements correctly.

You're right that it appears to be an integer in all of the examples. I made it a string because in the C++ version the `field-id` is required to be a string[^1]. It looks like in pyIceberg it has a type annotation as an `int` field, but pyspark is able to read files generated by this. It's possible we'll need to revisit this in the future and make it an int, or do a big overhaul of how custom attributes work in the C++ Avro implementation.

[^1]: This is actually an incompatibility with other versions of Avro. The original Iceberg schema that was generated by pySpark had integer field IDs. I changed them to strings and it worked.

Reply:
To clarify about the custom attribute thing:

An alternative to adding `element-id` to the array class would be to add custom attributes for arrays. In the C++ Avro implementation custom attributes are only supported for record types. If we added them for arrays element ID could just be a normal custom attribute instead of a special case.

This doesn't solve the typing problem. In C++ Avro custom attributes are a `map<string, string>`. To support integer attributes we would need to make custom attributes more generic. Digging through the commit history, it looks like that's how it used to work, but support for non-string attributes was removed a few years ago.

Reply:
>Since it's a value in the schema, my guess is that it's meant to identify the array element type, in the same way that field-id is used for record fields. So if you change the element type of an array without changing the name of an array, the system will be able to detect the schema evolution and parse the elements correctly.

Ah yeah, these docs pretty much confirm that, under "Field IDs": https://iceberg.apache.org/spec/#avro. BTW have you come across key-id and value-id? Does the C++ library support them well already?

AFAICT these all may as well have been some column-id, not quite sure why they opted to use different names per type.

Any idea how difficult it'd be to tweak the `field-id` to be an int? I guess I don't feel strongly enough or know enough about how this may bite us later, but if it's easy, might as well get it out of the way now?

Re: custom attributes, maybe a quick-ish hack would be to revive that feature but with a `map<string, int>` that'd be specific to custom numeric attributes? I guess that'd be pretty hard to upstream, but maybe not a huge priority atm

Reply:
I don't think that's the correct spec you linked. I think that's referring to the *Iceberg* schema, which is also written in JSON and looks a lot like an Avro schema.

Custom attributes weren't completely removed. In the original implementation they could be any type. That was quickly revised to be only strings.

- In [this commit](https://github.com/apache/avro/commit/61a1cdcd500356726d4995c19c8a4d5d89989b13) the type for the value is json::Entity 
- In [this commit](https://github.com/apache/avro/commit/864774b473202d11544cb869e9323de01c5ffe5c), 2 weeks later, it was changed to string. It looks like the reasoning was to avoid

Reply:
>I don't think that's the correct spec you linked. I think that's referring to the Iceberg schema, which is also written in JSON and looks a lot like an Avro schema.

I'm okay with that being the case, but just noting that what I'm referring to is also apparently an Avro schema.

>Iceberg struct, list, and map types identify nested types by ID. When writing data to Avro files, these IDs must be stored in the Avro schema to support ID-based column pruning.

I think this is saying that when writing data with the Avro format, the Avro schema needs to have `element-id` et al. Are you saying that this PR is required for all data types and is referring to the Iceberg schema? Might be helpful to include the sample schema so it's easier to point to exactly what you're referring to

Reply:
Oh yeah, it looks like you're right:
> Iceberg struct, list, and map types identify nested types by ID. When writing data to Avro files, these IDs must be stored in the Avro schema to support ID-based column pruning.

In all of the examples they give, the IDs are integers.

On the other hand, it looks like in the Java Avro implementation, custom attributes have to be strings: https://avro.apache.org/docs/1.6.2/api/csharp/classAvro_1_1Field.html#ac93ac61816cb1f110f6f84eb248f0643